File: sdk/villow-0.1.0/villow-0.1.0/src/villow/streaming.py
"""AG-UI-aligned streaming emit API for publisher agents (D-104, plan 11 Phase 6).

A publisher builds one turn's event stream with a fluent ``Stream`` and either streams it (``sse()``)
or hands the platform the event list (``events()``). The emitted events match the platform's canonical
contract (``apps/api/worklane_api/streaming/{events,parts}.py``) — same event-type names and part
schemas — so the platform ingests them with no translation. Publishers supply DATA; the platform owns
rendering (closed registry, D-106).

    s = Stream()
    s.reasoning("I see 47 PDFs — 9 are scans, 3 look like April duplicates.")
    s.text("Here's your **March books-ready package**.")
    s.table(columns=["merchant", "amount"], rows=[["ACME", "14302.11"]],
            reconciliation={"label": "Totals", "expected": "14302.11", "actual": "14302.11", "ok": True})
    s.require_approval(action="Write March-ledger.xlsx to your Drive")
    for chunk in s.sse():
        yield chunk

Non-streaming publishers can use ``Stream.from_batch(result)`` to turn a single result into a stream.
"""

from __future__ import annotations

import json
import time
import uuid
from typing import Any, Iterable

# Event-type constants (mirror worklane_api.streaming.events.EventType).
RUN_STARTED = "RUN_STARTED"
RUN_FINISHED = "RUN_FINISHED"
RUN_ERROR = "RUN_ERROR"
STEP_STARTED = "STEP_STARTED"
TEXT_MESSAGE_CHUNK = "TEXT_MESSAGE_CHUNK"
PART = "PART"
ARTIFACT_STAGED = "ARTIFACT_STAGED"
APPROVAL_REQUIRED = "APPROVAL_REQUIRED"
CLARIFICATION_REQUIRED = "CLARIFICATION_REQUIRED"
BUDGET_HOLD = "BUDGET_HOLD"
COST_UPDATE = "COST_UPDATE"
BUDGET_SETTLE = "BUDGET_SETTLE"


def _now_ms() -> int:
    return int(time.time() * 1000)


def _nonce() -> str:
    return uuid.uuid4().hex


def table_part(
    columns: list[Any],
    rows: list[Any],
    *,
    title: str | None = None,
    reconciliation: dict[str, Any] | None = None,
    exceptions: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Build a ``table_data`` part from friendly inputs (columns as names or {key,label,type})."""

    cols: list[dict[str, Any]] = []
    keys: list[str] = []
    for col in columns:
        if isinstance(col, dict):
            key = str(col.get("key") or col.get("label"))
            cols.append({"key": key, "label": col.get("label") or key, "type": col.get("type", "string")})
        else:
            key = str(col)
            cols.append({"key": key, "label": key})
        keys.append(key)
    out_rows: list[dict[str, Any]] = []
    for row in rows:
        if isinstance(row, dict):
            out_rows.append(row)
        elif isinstance(row, (list, tuple)):
            out_rows.append({(keys[i] if i < len(keys) else f"c{i}"): v for i, v in enumerate(row)})
        else:
            out_rows.append({"value": row})
    part: dict[str, Any] = {"type": "table_data", "columns": cols, "rows": out_rows}
    if title:
        part["title"] = title
    if reconciliation is not None:
        part["reconciliation"] = reconciliation
    if exceptions:
        part["exceptions"] = exceptions
    return part


def artifact_to_part(artifact: dict[str, Any]) -> dict[str, Any]:
    """Map a legacy typed-artifact dict ({artifact_type, title, type_payload}) onto a part."""

    atype = artifact.get("artifact_type") or artifact.get("type")
    payload = artifact.get("type_payload") or artifact.get("payload") or {}
    title = artifact.get("title")
    if atype == "table_data":
        return table_part(payload.get("columns") or [], payload.get("rows") or [], title=title,
                          reconciliation=payload.get("reconciliation"), exceptions=payload.get("exceptions"))
    if atype == "structured_fields":
        # The legacy artifact uses a dict of fields; the streaming part wants a list of {label, value}.
        fields = payload.get("fields")
        if isinstance(fields, dict):
            fields = [{"label": str(k), "value": v} for k, v in fields.items()]
        return {"type": "structured_fields", "fields": fields or []}
    if atype in {"file_set", "message_draft", "event_proposal"}:
        part = {"type": atype, **payload}
        if title and atype == "file_set":
            part["title"] = title
        return part
    return {"type": "generic", "data": {"artifact_type": atype, "title": title, **payload}, "original_type": atype}


def _progress_label(progress: dict[str, Any]) -> str | None:
    event_type = progress.get("event_type")
    if event_type == "milestone":
        milestone = progress.get("milestone") or {}
        return milestone.get("label") or milestone.get("id")
    if event_type == "item_completed":
        item = progress.get("item") or {}
        counts = progress.get("progress") or {}
        name = item.get("display_name") or item.get("item_id") or "item"
        if counts.get("total"):
            return f"{name} ({counts.get('current')}/{counts.get('total')})"
        return name
    return progress.get("label")


def _options_to_parts(options: Any) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for opt in options or []:
        out.append(opt if isinstance(opt, dict) else {"id": str(opt), "label": str(opt)})
    return out


def events_to_stream(emitted_events: list[dict[str, Any]], *, run_id: str | None = None) -> "Stream":
    """Adapt a publisher agent's SDK emissions into a canonical AG-UI Stream (plan 11 Phase 7).

    The migration path for existing SDK agents: they keep their rich ``ctx`` API (``report_progress`` /
    ``request_*_clarification`` / ``stage_artifact``) and this translates the recorded ``ctx.emitted_events``
    (``[{"kind", "payload"}]``) into the streaming contract — progress→step, clarification→clarify (with
    its decide_for_me preserved, rule #4), staged artifact→stage_artifact. Cost telemetry is billing
    input, not a renderable part, so it is intentionally not turned into a money part here.
    """

    stream = Stream(run_id)
    for event in emitted_events:
        kind = event.get("kind")
        payload = event.get("payload") or {}
        if kind == "status":
            progress = payload.get("progress")
            if isinstance(progress, dict):
                label = _progress_label(progress)
                if label:
                    stream.step(str(label))
        elif kind == "clarification_required":
            for question in payload.get("questions") or []:
                stream.clarify(
                    question=str(question.get("label") or question.get("question") or ""),
                    options=_options_to_parts(question.get("options")),
                    decide_for_me=question.get("decide_for_me"),
                    clarification_id=question.get("id"),
                )
        elif kind == "artifact_staged":
            artifact = payload.get("artifact")
            if artifact:
                stream.stage_artifact(artifact_to_part(artifact))
        elif kind == "sample_complete":
            for sample in payload.get("sample_artifacts") or []:
                stream.stage_artifact(artifact_to_part(sample))
    return stream.finish()


class Stream:
    """Builds one turn's canonical event stream. Auto-wraps RUN_STARTED/RUN_FINISHED."""

    def __init__(self, run_id: str | None = None) -> None:
        self.run_id = run_id or uuid.uuid4().hex
        self._events: list[dict[str, Any]] = []
        self._seq = 0
        self._started = False
        self._finished = False

    def _push(self, event_type: str, *, part: dict[str, Any] | None = None, data: dict[str, Any] | None = None) -> "Stream":
        event: dict[str, Any] = {"type": event_type, "seq": self._seq, "run_id": self.run_id, "ts": _now_ms(), "nonce": _nonce()}
        if part is not None:
            event["part"] = part
        event["data"] = data or {}
        self._events.append(event)
        self._seq += 1
        return self

    def _emit(self, event_type: str, *, part: dict[str, Any] | None = None, data: dict[str, Any] | None = None) -> "Stream":
        if not self._started:
            self._push(RUN_STARTED)
            self._started = True
        return self._push(event_type, part=part, data=data)

    # ----------------------------------------------------------------- content
    def reasoning(self, text: str) -> "Stream":
        return self._emit(PART, part={"type": "reasoning", "text": text})

    def text(self, text: str, *, role: str | None = None) -> "Stream":
        if role:
            return self._emit(TEXT_MESSAGE_CHUNK, data={"text": text, "role": role})
        return self._emit(PART, part={"type": "text", "text": text})

    def step(self, label: str) -> "Stream":
        return self._emit(STEP_STARTED, data={"name": label})

    def code(self, source: str, *, language: str | None = None, filename: str | None = None) -> "Stream":
        part: dict[str, Any] = {"type": "code", "source": source}
        if language:
            part["language"] = language
        if filename:
            part["filename"] = filename
        return self._emit(PART, part=part)

    def table(self, *, columns: list[Any], rows: list[Any], title: str | None = None,
              reconciliation: dict[str, Any] | None = None, exceptions: list[dict[str, Any]] | None = None) -> "Stream":
        return self._emit(ARTIFACT_STAGED, part=table_part(columns, rows, title=title, reconciliation=reconciliation, exceptions=exceptions))

    def file(self, *, id: str, name: str, mime: str, url: str, size_bytes: int | None = None) -> "Stream":
        ref: dict[str, Any] = {"id": id, "name": name, "mime": mime, "url": url}
        if size_bytes is not None:
            ref["size_bytes"] = size_bytes
        return self._emit(PART, part={"type": "file", "file": ref})

    def file_set(self, *, files: list[dict[str, Any]], title: str | None = None) -> "Stream":
        part: dict[str, Any] = {"type": "file_set", "files": files}
        if title:
            part["title"] = title
        return self._emit(ARTIFACT_STAGED, part=part)

    def stage_artifact(self, part: dict[str, Any]) -> "Stream":
        return self._emit(ARTIFACT_STAGED, part=part)

    # ----------------------------------------------------------------- trust
    def require_approval(self, *, action: str, summary: str | None = None, diff: dict[str, Any] | None = None,
                         target: dict[str, Any] | None = None, approval_id: str | None = None) -> "Stream":
        part: dict[str, Any] = {"type": "approval", "approval_id": approval_id or uuid.uuid4().hex, "action": action, "requires_approval": True}
        if summary:
            part["summary"] = summary
        if diff is not None:
            part["diff"] = diff
        if target is not None:
            part["target"] = target
        return self._emit(APPROVAL_REQUIRED, part=part)

    def clarify(self, *, question: str, options: list[dict[str, Any]] | None = None,
                decide_for_me: dict[str, Any] | None = None, allow_free_text: bool = False,
                clarification_id: str | None = None) -> "Stream":
        part: dict[str, Any] = {
            "type": "clarification",
            "clarification_id": clarification_id or uuid.uuid4().hex,
            "question": question,
            "options": options or [],
            "allow_free_text": allow_free_text,
            # rule #4: a clarification always carries a decide-for-me fallback.
            "decide_for_me": decide_for_me or {"label": "Decide for me"},
        }
        return self._emit(CLARIFICATION_REQUIRED, part=part)

    # ----------------------------------------------------------------- money
    def money_hold(self, *, typical: int | None = None, ceiling: int | None = None, hold: int | None = None) -> "Stream":
        part: dict[str, Any] = {"type": "money", "kind": "hold"}
        if typical is not None:
            part["typical_credits"] = typical
        if ceiling is not None:
            part["ceiling_credits"] = ceiling
        if hold is not None:
            part["hold_credits"] = hold
        return self._emit(BUDGET_HOLD, part=part)

    def money_settle(self, *, captured: int, hold: int | None = None, stats: dict[str, Any] | None = None) -> "Stream":
        part: dict[str, Any] = {"type": "money", "kind": "settle", "captured_credits": captured}
        if hold is not None:
            part["hold_credits"] = hold
        if stats is not None:
            part["stats"] = stats
        return self._emit(BUDGET_SETTLE, part=part)

    # ----------------------------------------------------------------- lifecycle
    def finish(self) -> "Stream":
        if not self._started:
            self._push(RUN_STARTED)
            self._started = True
        if not self._finished:
            self._push(RUN_FINISHED)
            self._finished = True
        return self

    def error(self, message: str) -> "Stream":
        if not self._started:
            self._push(RUN_STARTED)
            self._started = True
        if not self._finished:
            self._push(RUN_ERROR, data={"message": message})
            self._finished = True
        return self

    def events(self) -> list[dict[str, Any]]:
        if not self._finished:
            self.finish()
        return list(self._events)

    def sse(self) -> Iterable[str]:
        for event in self.events():
            yield f"id: {event['seq']}\nevent: {event['type']}\ndata: {json.dumps(event, separators=(',', ':'))}\n\n"

    @classmethod
    def from_batch(cls, result: Any, run_id: str | None = None) -> "Stream":
        """Batch→stream fallback (P6.3): a non-streaming publisher's single result → a part stream."""

        stream = cls(run_id)
        if isinstance(result, dict) and isinstance(result.get("parts"), list):
            parts = list(result["parts"])
        elif isinstance(result, dict) and result.get("artifact"):
            parts = [artifact_to_part(result["artifact"])]
        elif isinstance(result, dict) and isinstance(result.get("artifacts"), list):
            parts = [artifact_to_part(a) for a in result["artifacts"]]
        elif isinstance(result, dict):
            parts = [{"type": "generic", "data": dict(result)}]
        else:
            parts = [{"type": "generic", "data": {"value": result}}]
        for part in parts:
            stream.stage_artifact(part)
        return stream.finish()

File: sdk/villow-sdk-0.1.0/package/src/streaming.ts
// AG-UI-aligned streaming emit API for publisher agents (D-104, plan 11 Phase 6).
// Node parity with sdk/python/src/villow/streaming.py: same event-type names and part schemas (wire
// fields stay snake_case), so the platform ingests both SDKs identically. Publishers supply DATA; the
// platform owns rendering (closed registry, D-106).

import { randomUUID } from 'node:crypto';

export type Part = { type: string; [key: string]: unknown };
export type StreamEvent = {
  type: string;
  seq: number;
  run_id: string;
  ts: number;
  nonce: string;
  part?: Part;
  data: Record<string, unknown>;
};

export const EVENT = {
  RUN_STARTED: 'RUN_STARTED',
  RUN_FINISHED: 'RUN_FINISHED',
  RUN_ERROR: 'RUN_ERROR',
  STEP_STARTED: 'STEP_STARTED',
  TEXT_MESSAGE_CHUNK: 'TEXT_MESSAGE_CHUNK',
  PART: 'PART',
  ARTIFACT_STAGED: 'ARTIFACT_STAGED',
  APPROVAL_REQUIRED: 'APPROVAL_REQUIRED',
  CLARIFICATION_REQUIRED: 'CLARIFICATION_REQUIRED',
  BUDGET_HOLD: 'BUDGET_HOLD',
  COST_UPDATE: 'COST_UPDATE',
  BUDGET_SETTLE: 'BUDGET_SETTLE',
} as const;

type Column = string | { key?: string; label?: string; type?: string };

export function tablePart(
  columns: Column[],
  rows: Array<unknown[] | Record<string, unknown>>,
  opts: { title?: string; reconciliation?: Record<string, unknown>; exceptions?: Array<Record<string, unknown>> } = {},
): Part {
  const cols: Array<{ key: string; label: string; type?: string }> = [];
  const keys: string[] = [];
  for (const col of columns) {
    if (typeof col === 'object') {
      const key = String(col.key ?? col.label);
      cols.push({ key, label: col.label ?? key, type: col.type ?? 'string' });
      keys.push(key);
    } else {
      cols.push({ key: String(col), label: String(col) });
      keys.push(String(col));
    }
  }
  const outRows = rows.map((row) => {
    if (Array.isArray(row)) {
      const obj: Record<string, unknown> = {};
      row.forEach((v, i) => {
        obj[keys[i] ?? `c${i}`] = v;
      });
      return obj;
    }
    return row;
  });
  const part: Part = { type: 'table_data', columns: cols, rows: outRows };
  if (opts.title) part.title = opts.title;
  if (opts.reconciliation) part.reconciliation = opts.reconciliation;
  if (opts.exceptions && opts.exceptions.length) part.exceptions = opts.exceptions;
  return part;
}

export function artifactToPart(artifact: Record<string, unknown>): Part {
  const atype = (artifact.artifact_type ?? artifact.type) as string | undefined;
  const payload = (artifact.type_payload ?? artifact.payload ?? {}) as Record<string, unknown>;
  const title = artifact.title as string | undefined;
  if (atype === 'table_data') {
    return tablePart((payload.columns as Column[]) ?? [], (payload.rows as unknown[][]) ?? [], {
      title,
      reconciliation: payload.reconciliation as Record<string, unknown> | undefined,
      exceptions: payload.exceptions as Array<Record<string, unknown>> | undefined,
    });
  }
  if (atype === 'structured_fields') {
    // The legacy artifact uses a dict of fields; the streaming part wants a list of {label, value}.
    let fields = payload.fields as unknown;
    if (fields && typeof fields === 'object' && !Array.isArray(fields)) {
      fields = Object.entries(fields as Record<string, unknown>).map(([k, v]) => ({ label: k, value: v }));
    }
    return { type: 'structured_fields', fields: (fields as unknown[]) ?? [] };
  }
  if (atype && ['file_set', 'message_draft', 'event_proposal'].includes(atype)) {
    const part: Part = { type: atype, ...payload };
    if (title && atype === 'file_set') part.title = title;
    return part;
  }
  return { type: 'generic', data: { artifact_type: atype, title, ...payload }, original_type: atype };
}

function progressLabel(progress: Record<string, unknown>): string | undefined {
  const eventType = progress.event_type;
  if (eventType === 'milestone') {
    const m = (progress.milestone ?? {}) as Record<string, unknown>;
    return (m.label ?? m.id) as string | undefined;
  }
  if (eventType === 'item_completed') {
    const item = (progress.item ?? {}) as Record<string, unknown>;
    const counts = (progress.progress ?? {}) as Record<string, unknown>;
    const name = (item.display_name ?? item.item_id ?? 'item') as string;
    return counts.total ? `${name} (${counts.current}/${counts.total})` : name;
  }
  return progress.label as string | undefined;
}

function optionsToParts(options: unknown): Array<Record<string, unknown>> {
  if (!Array.isArray(options)) return [];
  return options.map((o) => (o && typeof o === 'object' ? (o as Record<string, unknown>) : { id: String(o), label: String(o) }));
}

/**
 * Adapt a publisher agent's SDK emissions ([{kind, payload}]) into a canonical AG-UI Stream
 * (plan 11 Phase 7) — the migration path for existing SDK agents. Parity with
 * villow.streaming.events_to_stream (Python).
 */
export function eventsToStream(emittedEvents: Array<{ kind?: string; payload?: Record<string, unknown> }>, runId?: string): Stream {
  const s = new Stream(runId);
  for (const ev of emittedEvents) {
    const payload = ev.payload ?? {};
    if (ev.kind === 'status') {
      const progress = payload.progress as Record<string, unknown> | undefined;
      if (progress && typeof progress === 'object') {
        const label = progressLabel(progress);
        if (label) s.step(label);
      }
    } else if (ev.kind === 'clarification_required') {
      for (const q of (payload.questions as Array<Record<string, unknown>>) ?? []) {
        s.clarify({
          question: String(q.label ?? q.question ?? ''),
          options: optionsToParts(q.options),
          decideForMe: q.decide_for_me as Record<string, unknown> | undefined,
          clarificationId: q.id as string | undefined,
        });
      }
    } else if (ev.kind === 'artifact_staged') {
      if (payload.artifact) s.stageArtifact(artifactToPart(payload.artifact as Record<string, unknown>));
    } else if (ev.kind === 'sample_complete') {
      for (const a of (payload.sample_artifacts as Array<Record<string, unknown>>) ?? []) s.stageArtifact(artifactToPart(a));
    }
  }
  return s.finish();
}

export class Stream {
  readonly runId: string;
  private events_: StreamEvent[] = [];
  private seq = 0;
  private started = false;
  private finished = false;

  constructor(runId?: string) {
    this.runId = runId ?? randomUUID().replace(/-/g, '');
  }

  private push(type: string, part?: Part, data?: Record<string, unknown>): this {
    const event: StreamEvent = { type, seq: this.seq, run_id: this.runId, ts: Date.now(), nonce: randomUUID().replace(/-/g, ''), data: data ?? {} };
    if (part !== undefined) event.part = part;
    this.events_.push(event);
    this.seq += 1;
    return this;
  }

  private emit(type: string, part?: Part, data?: Record<string, unknown>): this {
    if (!this.started) {
      this.push(EVENT.RUN_STARTED);
      this.started = true;
    }
    return this.push(type, part, data);
  }

  // content
  reasoning(text: string): this {
    return this.emit(EVENT.PART, { type: 'reasoning', text });
  }
  text(text: string, role?: string): this {
    if (role) return this.emit(EVENT.TEXT_MESSAGE_CHUNK, undefined, { text, role });
    return this.emit(EVENT.PART, { type: 'text', text });
  }
  step(label: string): this {
    return this.emit(EVENT.STEP_STARTED, undefined, { name: label });
  }
  code(source: string, opts: { language?: string; filename?: string } = {}): this {
    const part: Part = { type: 'code', source };
    if (opts.language) part.language = opts.language;
    if (opts.filename) part.filename = opts.filename;
    return this.emit(EVENT.PART, part);
  }
  table(opts: { columns: Column[]; rows: Array<unknown[] | Record<string, unknown>>; title?: string; reconciliation?: Record<string, unknown>; exceptions?: Array<Record<string, unknown>> }): this {
    return this.emit(EVENT.ARTIFACT_STAGED, tablePart(opts.columns, opts.rows, opts));
  }
  file(opts: { id: string; name: string; mime: string; url: string; size_bytes?: number }): this {
    const ref: Record<string, unknown> = { id: opts.id, name: opts.name, mime: opts.mime, url: opts.url };
    if (opts.size_bytes !== undefined) ref.size_bytes = opts.size_bytes;
    return this.emit(EVENT.PART, { type: 'file', file: ref });
  }
  fileSet(opts: { files: Array<Record<string, unknown>>; title?: string }): this {
    const part: Part = { type: 'file_set', files: opts.files };
    if (opts.title) part.title = opts.title;
    return this.emit(EVENT.ARTIFACT_STAGED, part);
  }
  stageArtifact(part: Part): this {
    return this.emit(EVENT.ARTIFACT_STAGED, part);
  }

  // trust
  requireApproval(opts: { action: string; summary?: string; diff?: Record<string, unknown>; target?: Record<string, unknown>; approvalId?: string }): this {
    const part: Part = { type: 'approval', approval_id: opts.approvalId ?? randomUUID().replace(/-/g, ''), action: opts.action, requires_approval: true };
    if (opts.summary) part.summary = opts.summary;
    if (opts.diff !== undefined) part.diff = opts.diff;
    if (opts.target !== undefined) part.target = opts.target;
    return this.emit(EVENT.APPROVAL_REQUIRED, part);
  }
  clarify(opts: { question: string; options?: Array<Record<string, unknown>>; decideForMe?: Record<string, unknown>; allowFreeText?: boolean; clarificationId?: string }): this {
    const part: Part = {
      type: 'clarification',
      clarification_id: opts.clarificationId ?? randomUUID().replace(/-/g, ''),
      question: opts.question,
      options: opts.options ?? [],
      allow_free_text: opts.allowFreeText ?? false,
      // rule #4: a clarification always carries a decide-for-me fallback.
      decide_for_me: opts.decideForMe ?? { label: 'Decide for me' },
    };
    return this.emit(EVENT.CLARIFICATION_REQUIRED, part);
  }

  // money
  moneyHold(opts: { typical?: number; ceiling?: number; hold?: number } = {}): this {
    const part: Part = { type: 'money', kind: 'hold' };
    if (opts.typical !== undefined) part.typical_credits = opts.typical;
    if (opts.ceiling !== undefined) part.ceiling_credits = opts.ceiling;
    if (opts.hold !== undefined) part.hold_credits = opts.hold;
    return this.emit(EVENT.BUDGET_HOLD, part);
  }
  moneySettle(opts: { captured: number; hold?: number; stats?: Record<string, unknown> }): this {
    const part: Part = { type: 'money', kind: 'settle', captured_credits: opts.captured };
    if (opts.hold !== undefined) part.hold_credits = opts.hold;
    if (opts.stats !== undefined) part.stats = opts.stats;
    return this.emit(EVENT.BUDGET_SETTLE, part);
  }

  // lifecycle
  finish(): this {
    if (!this.started) {
      this.push(EVENT.RUN_STARTED);
      this.started = true;
    }
    if (!this.finished) {
      this.push(EVENT.RUN_FINISHED);
      this.finished = true;
    }
    return this;
  }
  error(message: string): this {
    if (!this.started) {
      this.push(EVENT.RUN_STARTED);
      this.started = true;
    }
    if (!this.finished) {
      this.push(EVENT.RUN_ERROR, undefined, { message });
      this.finished = true;
    }
    return this;
  }

  events(): StreamEvent[] {
    if (!this.finished) this.finish();
    return [...this.events_];
  }

  *sse(): IterableIterator<string> {
    for (const event of this.events()) {
      yield `id: ${event.seq}\nevent: ${event.type}\ndata: ${JSON.stringify(event)}\n\n`;
    }
  }

  static fromBatch(result: unknown, runId?: string): Stream {
    const stream = new Stream(runId);
    let parts: Part[];
    const r = result as Record<string, unknown>;
    if (r && typeof r === 'object' && Array.isArray(r.parts)) parts = r.parts as Part[];
    else if (r && typeof r === 'object' && r.artifact) parts = [artifactToPart(r.artifact as Record<string, unknown>)];
    else if (r && typeof r === 'object' && Array.isArray(r.artifacts)) parts = (r.artifacts as Array<Record<string, unknown>>).map(artifactToPart);
    else if (r && typeof r === 'object') parts = [{ type: 'generic', data: { ...r } }];
    else parts = [{ type: 'generic', data: { value: result } }];
    for (const part of parts) stream.stageArtifact(part);
    return stream.finish();
  }
}
